### ISSUE TITLE
OCI Store resolver returns plain descriptor

### ISSUE THREAD

**Issue by carabasdaniel** [oras-project/oras-go#457]
When resolving a descriptor from the OCI store, the returned descriptor is stripped down to contain only the media type, digest and size: https://github.com/oras-project/oras-go/blob/96a37c2b359ac1305f70dc31b28c789688d77d0f/content/oci/oci.go#L185

Is there any particular reason to not have a [complete](https://github.com/opencontainers/image-spec/blob/819aa940cae7c067a8bf89b1745d3255ddaaba1d/specs-go/v1/descriptor.go) descriptor returned ?

**Comment by Wwwsylvia** [comment]
Returning the complete descriptor could be tricky in some scenarios.

I would like to take the chance to discuss this use case:
Suppose there are two descriptors pointing to the same manifest with different tags (`v1.0` and `latest`) in the `index.json`:

```json
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.index.v1+json",
    "manifests": [
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "org.opencontainers.image.ref.name": "v1.0",
          "foo":"bar"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "org.opencontainers.image.ref.name": "latest",
          "foo":"bar"
        }
      }
    ]
  }
```
If we resolve the descriptor by its digest:
```go
desc, err := oci.Resolve(ctx, "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f")
```
What should be the result of `oci.Resolve()`?

Should it be the descriptor tagged by `v1.0`?
```json
     {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "org.opencontainers.image.ref.name": "v1.0",
          "foo":"bar"
        }
      },
```
Or the descriptor tagged by `latest`?
```json
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "org.opencontainers.image.ref.name": "latest",
          "foo":"bar"
        }
      }
```
Or should we strip the annotation `org.opencontainers.image.ref.name` and keep others?
```json
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "foo":"bar"
        }
      }
```

**Comment by carabasdaniel** [comment]
Hi @Wwwsylvia,
Thanks for the quick reply. 

Ideally all annotations should be kept and the resolver should return an array of descriptors for the requested digest. 

To keep things simple it might make sense to strip the `org.opencontainers.image.ref.name` from annotations and keep everything else.

**Comment by gertd** [comment]
@Wwwsylvia it is my understanding that in the current OCI model, it is a legal and common case that a `sha256` can resolve to 1 to N number of tags. Not supporting that would introduce backward compatibility backlash in my opinion which is undesirable. Am I missing something?

**Comment by Wwwsylvia** [comment]
@gertd Yes a digest can be associated with multiple tags, and all the tags are saved in `index.json`.

> a sha256 can resolve to 1 to N number of tags. 

Do you mean we should return a list of descriptors for `Resolve(dgst)` ? Or am I understanding it wrong?

> Not supporting that would introduce backward compatibility backlash

Could you elaborate a bit more on "backward compatibility"?

**Comment by Wwwsylvia** [comment]
Maybe we can add a function `ResolveN()`, which returns multiple descriptors for a digest reference and returns one descriptor for a tag reference.

```go
func (s *Store) ResolveN(ctx context.Context, reference string) ([]ocispec.Descriptor, error)
```

Using the above example, if we call

```go
descs, err := oci.ResolveN(ctx, "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f")
```
The result would be
```json
     {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "org.opencontainers.image.ref.name": "v1.0",
          "foo":"bar"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "org.opencontainers.image.ref.name": "latest",
          "foo":"bar"
        }
    }
```

And if we call
```go
descs, err := oci.ResolveN(ctx, "latest")
```

The result would be

```json
     {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "org.opencontainers.image.ref.name": "latest",
          "foo":"bar"
        }
      }
```

**Comment by Wwwsylvia** [comment]
For the original `Resolve()` function, how about returning a complete descriptor for a tag reference and returning a descriptor without `org.opencontainers.image.ref.name` for a digest reference? 🤔

Code:
```go
desc, err := oci.Resolve(ctx, "latest")
```
Result:
```json
    {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "foo":"bar",
          "org.opencontainers.image.ref.name": "latest",
        }
      }
```
---
Code:
```go
desc, err := oci.Resolve(ctx, "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f")
```
Result:
```json
    {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 7143,
        "digest": "sha256:e692418e4cbaf90ca69d05a66403747baa33ee08806650b51fab815ad7fc331f",
        "annotations": {
          "foo":"bar",
        }
      }
```

**Comment by Wwwsylvia** [comment]
@shizhMSFT What do you think?

**Comment by shizhMSFT** [comment]
`Resolve()` digest is ambiguous if the manifest has multiple tags along with annotations. In that case, we can have `ResolveN()` to reduce ambiguity.

My suggestion would be
- If reference is a digest,
  - `Resolve()` returns a plain descriptor
  - `ResolveN()` return an array of descriptors with annotations
- If reference is a tag,
  - `Resolve()` return a descriptor with annotations
  - `ResolveN()` works the same as `Resolve()` but return an array

**Comment by qweeah** [comment]
This is a breaking change which introduces inconsistent behavior of resolver between remote and OCI store.

**Comment by Wwwsylvia** [comment]
> This is a breaking change which introduces inconsistent behavior of resolver between remote and OCI store.

@qweeah Could you elaborate a bit on this?

**Comment by qweeah** [comment]
Well remote registry manifest resolver always returns a descriptor without annotation.
https://github.com/oras-project/oras-go/blob/e8225cb1e125bd4c13d6b586ae6d862050c3fae2/registry/remote/repository.go#L1372-L1380

With merging of https://github.com/oras-project/oras-go/commit/e8225cb1e125bd4c13d6b586ae6d862050c3fae2, resolve a tagged manifest via tag will always a descriptor with at least one `org.opencontainers.image.ref.name` annotation.

Noticed this because of the [E2E failure](https://github.com/oras-project/oras/actions/runs/4777187901/jobs/8492719377#step:6:1385) in ORAS CLI:
I copied a tagged manifest from remote registry to local OCI layout via tag, resolving the manifest via tag and compare it to the resolved descriptor in registry. The content doesn't match so the E2E fails.

**Comment by qweeah** [comment]
This might not be a critical change since oras-go doesn't make any promise to guarantee that the resolved descriptor from registry and OCI layout will be identical. But IMHO it's better to bring it to up for discussion, like whether it's a breaking change and shall we pick it out from upcoming V2 minor releases.

**Comment by Wwwsylvia** [comment]
> This might not be a critical change since oras-go doesn't make any promise to guarantee that the resolved descriptor from registry and OCI layout will be identical.

Yeah there is no a guarantee that `OCI.Resolve()` and `Repository.Resolve()` will return exactly the same result.
For OCI Target, the manifest descriptor is resolved from the records in `index.json`. It's intuitive that the resolved descriptor matches the record in `index.json`.
But there is no such `index.json` for remote repositories, and the descriptor is resolved from the HTTP response of the remote server.

> But IMHO it's better to bring it to up for discussion, like whether it's a breaking change and shall we pick it out from upcoming V2 minor releases.

I personally don't think it's a breaking change for `oras-go`, as it's adding things (annotation, etc.) to the original result (plain descriptor), not changing or removing existing stuffs. 
But it could be a breaking change from the ORAS CLI point of view since the output format changed. 🤔
IMHO we can update the ORAS E2E tests to accommodate this change.

@shizhMSFT Any thoughts?

**Comment by shizhMSFT** [comment]
In general, `Target.Resolve()` can return any extra stuffs other than the digest, size, media type. Therefore, we should use [content.Equal()](https://pkg.go.dev/oras.land/oras-go/v2/content#Equal) to compare the returned descriptors instead of `reflect.DeepEqual()`.

If ORAS CLI E2E tests has more context on the `Target` instead of a generic `Target`, the E2E test can check the returned descriptor in deep accordingly.

**Comment by shizhMSFT** [comment]
Closed by #468

**Comment by Wwwsylvia** [comment]
> `Resolve()` digest is ambiguous if the manifest has multiple tags along with annotations. In that case, we can have `ResolveN()` to reduce ambiguity.
> 
> My suggestion would be
> 
> * If reference is a digest,
>   
>   * `Resolve()` returns a plain descriptor
>   * `ResolveN()` return an array of descriptors with annotations
> * If reference is a tag,
>   
>   * `Resolve()` return a descriptor with annotations
>   * `ResolveN()` works the same as `Resolve()` but return an array

The `ResolveN` function has not been implemented yet. Do we need to raise another issue to track this feature request?
